<div style="
  background: linear-gradient(135deg, #ff9a9e, #fad0c4, #fbc2eb);
  padding: 28px;
  border-radius: 24px;
  text-align: center;
  color: #3a2c4a;
  box-shadow: 0 8px 20px rgba(0,0,0,0.12);
">

  <h1 style="font-size: 42px; margin-bottom: 8px;">
    🌈 DaneshChat AI Adventure 🚀
  </h1>

  <h2 style="font-size: 24px; margin-top: 0;">
    Build Smart Models for Spam & Message Loss Detection
  </h2>

  <p style="font-size: 18px;">
    Welcome, young AI engineer! Today you will train machines to understand messages better.
  </p>

</div>

### Imports

In [1]:
import numpy as np
import pandas as pd

### Load the dataset

In [2]:
df = pd.read_csv("daneshchat_logs.csv")

print(df.head())

   message_size_kb  transfer_steps  connection_quality  has_urgent_keyword  \
0            49.73              10               0.954                   0   
1            60.27               4               0.369                   0   
2            77.18              17               0.025                   0   
3            21.13               8               0.334                   1   
4            39.09               2               0.664                   0   

   has_link  is_spam  is_lost  
0         0        0        0  
1         0        0        1  
2         0        0        1  
3         0        1        1  
4         0        0        0  


### Quick data check

In [3]:
# TODO: Print the shape of the dataset
# TODO: Print the column names
# TODO: Check missing values
# TODO: Show basic statistics using describe()

print(df.shape)

print(df.columns)

print(df.isnull().sum())

print(df.describe())

(6000, 7)
Index(['message_size_kb', 'transfer_steps', 'connection_quality',
       'has_urgent_keyword', 'has_link', 'is_spam', 'is_lost'],
      dtype='str')
message_size_kb       0
transfer_steps        0
connection_quality    0
has_urgent_keyword    0
has_link              0
is_spam               0
is_lost               0
dtype: int64
       message_size_kb  transfer_steps  connection_quality  \
count      6000.000000     6000.000000         6000.000000   
mean         65.794645        8.684667            0.595544   
std         121.959485        6.041494            0.266757   
min           1.000000        1.000000            0.000000   
25%          23.342500        4.000000            0.399750   
50%          40.830000        7.000000            0.644000   
75%          67.992500       12.000000            0.811250   
max        1199.570000       48.000000            1.000000   

       has_urgent_keyword     has_link      is_spam      is_lost  
count         6000.000000  6000.00

### Clean the dataset

In [4]:
# TODO: Remove duplicated rows
# TODO: Convert columns to numeric if needed
# TODO: Remove rows with missing values
# TODO: Keep only valid connection_quality values between 0 and 1
# TODO: Keep only positive message_size_kb and transfer_steps

df = df.drop_duplicates()

df = df.apply(pd.to_numeric,errors = 'coerce')

df = df.dropna()

df = df[(df['connection_quality'] >= 0) & (df['connection_quality'] <= 1)]

df = df[(df['message_size_kb'] > 0) & (df['transfer_steps'] > 0)]

print(df.shape)

(6000, 7)


<div style="
  background: linear-gradient(135deg, #74c69d, #48cae4);
  padding: 26px;
  border-radius: 22px;
  margin-top: 28px;
  border: 4px solid #2d6a4f;
  box-shadow: 0 8px 22px rgba(0, 0, 0, 0.18);
  color: #081c15;
">

  <h1 style="
    color: #081c15;
    background-color: rgba(255, 255, 255, 0.65);
    padding: 12px 16px;
    border-radius: 14px;
    margin-top: 0;
  ">
    🌳 Part A: Build a Decision Tree from Scratch
  </h1>

  <p style="
    font-size: 17px;
    line-height: 1.7;
    color: #133f30;
    font-weight: 500;
  ">
    In this part, you will build your own Decision Tree classifier.
    You may use Python, NumPy, and Pandas, but not a ready-made Decision Tree model.
  </p>

  <pre style="
    background-color: #0f4331;
    color: #d8f3dc;
    padding: 16px;
    border-radius: 14px;
    border: 3px solid #95d5b2;
    font-size: 15px;
    line-height: 1.6;
    overflow-x: auto;
  ">Is connection_quality <= 0.45?

├── Yes: maybe lost

└── No: maybe delivered</pre>

</div>

### Create a Node class

In [5]:
class Node:
    def __init__(self, feature=None, threshold=None, left=None, right=None, value=None):
        # TODO: Store the feature name used for splitting
        # TODO: Store the threshold value
        # TODO: Store the left child
        # TODO: Store the right child
        # TODO: Store the final prediction value for leaf nodes

        self.feature = feature
        self.threshold = threshold
        self.left = left
        self.right = right
        self.value = value

    def is_leaf(self):
        # TODO: Return True if this node is a leaf
        if self.value is not None:
            return True

        return False

### Decision Tree Classifier from Scratch

In [6]:
class DecisionTreeClassifierScratch:
    def __init__(self, max_depth=5, min_samples_split=20, min_samples_leaf=10):
        # TODO: Store max_depth
        # TODO: Store min_samples_split
        # TODO: Store min_samples_leaf
        # TODO: Create a variable named root and set it to None
        self.max_depth = max_depth
        self.min_samples_split = min_samples_split
        self.min_samples_leaf = min_samples_leaf
        self.root = None

    def entropy(self, y):
        # TODO: Calculate entropy using class probabilities
        # Hint:
        # values, counts = np.unique(y, return_counts=True)
        # probabilities = counts / len(y)

        if len(y) == 0:
            return 0
        
        values, counts = np.unique(y,return_counts=True)

        probabilities = counts / len(y)

        return -np.sum(probabilities * np.log2(probabilities))

    def information_gain(self, y, y_left, y_right):
        # TODO: Calculate parent entropy
        # TODO: Calculate weighted child entropy
        # TODO: Return information gain
        
        H_s = self.entropy(y)

        H_l = self.entropy(y_left)

        H_r = self.entropy(y_right)

        weight_l = len(y_left) / len(y)
        weight_r = len(y_right) / len(y)

        IG = H_s - weight_r * H_r - weight_l * H_l

        return IG

    def majority_class(self, y):
        # TODO: Return the most common class in y
        if len(y) == 0:
            return 0
        

        values, counts = np.unique(y,return_counts=True)

        best_idx = np.argmax(counts)

        return values[best_idx]
    
    
    def find_best_split(self, X, y):
        # TODO: Try each feature
        # TODO: Try possible thresholds for each feature
        # TODO: Split y into left and right groups
        # TODO: Calculate information gain
        # TODO: Return best feature and best threshold
        
        best_ig = -1

        best_feature = None

        best_threshold = None

        samples, features = X.shape

        for f in range(features):

            X_c = X[:,f]

            thresholds = np.unique(X_c)

            for t in thresholds:
                left = X_c <= t
                right = X_c > t

                y_left = y[left] 
                y_right = y[right]

                if len(y_left) < self.min_samples_leaf or len(y_right) < self.min_samples_leaf:
                    continue

                ig = self.information_gain(y,y_left,y_right)

                if ig > best_ig:
                    best_ig = ig
                    best_feature = f
                    best_threshold = t

        return best_feature,best_threshold

        
    def build_tree(self, X, y, depth=0):
        # TODO: Check stopping conditions
        # TODO: Find the best split
        # TODO: Build left and right child nodes recursively
        # TODO: Return a Node

        samples, features = X.shape

        if depth >= self.max_depth or samples < self.min_samples_split or len(np.unique(y)) == 1:
            leaf_value = self.majority_class(y)

            return Node(value= leaf_value)

        best_feature, best_threshold = self.find_best_split(X,y)
        
        if best_feature is None:
            leaf_value = self.majority_class(y)
            return Node(value=leaf_value)
        left = X[:,best_feature] <= best_threshold
        right = X[:,best_feature] > best_threshold
        
        left_child = self.build_tree(X[left],y[left],depth + 1)
        right_child = self.build_tree(X[right],y[right],depth + 1)

        return Node(left=left_child,right=right_child,feature=best_feature,threshold=best_threshold)

    def fit(self, X, y):
        # TODO: Build the tree and store it in self.root
         self.root = self.build_tree(X,y)

        
    def predict_one(self, row):
        # TODO: Start from the root
        # TODO: Move left or right until reaching a leaf
        # TODO: Return the leaf value

        node = self.root

        while not node.is_leaf():
            if row[node.feature] <= node.threshold:
                node = node.left
            else:   
                node = node.right

        return node.value 

    def predict(self, X):
        # TODO: Predict each row in X

        predictions = np.zeros(len(X))
        i = 0
        for row in X:
            temp = self.predict_one(row)

            predictions[i] = temp
            i += 1
        return predictions

### Prepare data for spam prediction

In [7]:
# TODO: Set y_spam to df["is_spam"]
# TODO: Set X_spam by dropping both "is_spam" and "is_lost"
# TODO: Split X_spam and y_spam into train and validation sets
y_spam = df["is_spam"]

X_spam = df.drop(columns=["is_lost","is_spam"])

np.random.seed(2006)

shuffled_indices = np.random.permutation(len(X_spam))

test_size = 0.2

split = int(len(X_spam) * (1 - test_size))

train_idx = shuffled_indices[:split]

val_idx = shuffled_indices[split:]

X_train = X_spam.iloc[train_idx]
y_train = y_spam.iloc[train_idx]

X_val = X_spam.iloc[val_idx]
y_val = y_spam.iloc[val_idx]


### Train your scratch Decision Tree for spam

In [8]:
# TODO: Create a DecisionTreeClassifierScratch object
# TODO: Fit it on X_train_spam and y_train_spam
# TODO: Predict on X_val_spam

X_train_np = X_train.to_numpy()
y_train_np = y_train.to_numpy()

X_val_np = X_val.to_numpy()

tree = DecisionTreeClassifierScratch(max_depth=5)

tree.fit(X_train_np,y_train_np)

pred_spam = tree.predict(X_val_np)

### Create an evaluation function

In [9]:
def evaluate_model(y_true, y_pred, model_name):
    # TODO: Calculate accuracy
    # TODO: Calculate precision
    # TODO: Calculate recall
    # TODO: Calculate F1-score
    # TODO: Print the results in a nice format
    
    tp = np.sum((y_true == 1) & (y_pred == 1))
    tn = np.sum((y_true == 0) & (y_pred == 0))
    fp = np.sum((y_true == 0) & (y_pred == 1))
    fn = np.sum((y_true == 1) & (y_pred == 0))


    accuracy = (tp + tn) / (tp + tn + fp + fn)

    precision = tp / (tp + fp) 

    recall = tp / (fn + tp)

    F1_score = 2 * (precision * recall) / (precision + recall)

    print(f" Performance Metrics: {model_name}")
    print(f"------------------------------------------")
    print(f"Accuracy:  {accuracy:.4f}")
    print(f"Precision: {precision:.4f}")
    print(f"Recall:    {recall:.4f}")
    print(f"F1-Score:  {F1_score:.4f}")
    print(f"------------------------------------------")
    print(f"Confusion Breakdown ====> TP: {tp} | TN: {tn} | FP: {fp} | FN: {fn}")




In [10]:
evaluate_model(y_val,pred_spam,"custom tree")

 Performance Metrics: custom tree
------------------------------------------
Accuracy:  0.6975
Precision: 0.7279
Recall:    0.4422
F1-Score:  0.5502
------------------------------------------
Confusion Breakdown ====> TP: 222 | TN: 615 | FP: 83 | FN: 280


***Metrics for is_lost:***

 Performance Metrics: custom tree
------------------------------------------

Accuracy:  0.6483

Precision: 0.6649

Recall:    0.6258

F1-Score:  0.6448

------------------------------------------

Confusion Breakdown ====> TP: 383 | TN: 395 | FP: 193 | FN: 229

***target: is_lost***

*Accuracy:*

The model correctly predicts whether the message is lost or safely delivered about 65% of the time.

*Precision:*

When the model determine the message is lost we can say it is 66% correct.

*Recall:*

It approximately catches the 63% of the lost messages.

*F1-score:*

The value is roughly 64.5%, Because this score is very close to our overall accuracy, it shows that our model maintains a relatively healthy balance between catching lost messages and avoiding false alarms, without being heavily biased toward one specific class.

***Confusion Matrix:***

True Positives (383): The model correctly identified 383 messages that were going to get lost.

True Negatives (395): The model correctly identified 395 messages that safely reached their destination.

False Positives (193): The model raised a false alarm. It flagged 193 healthy messages as "lost" even though they actually arrived safely.

False Negatives (229): The model had a blind spot. It completely missed 229 lost messages, assuming they were safely delivered when they were actually dropped by the network.





<div style="
  background: linear-gradient(135deg, #e8b7e8, #ffb6f4);
  padding: 24px;
  border-radius: 22px;
  margin-top: 28px;
  border: 4px solid #481344;
  color: #2b0a2a;
">

  <h1 style="color: #2b0a2a;">⚡ Part B: Fast Spam Filtering with Scikit-learn</h1>

  <p style="font-size: 17px; line-height: 1.7; color: #2b0a2a;">
    Now you can use ready-made tools from Scikit-learn.
    In this part, focus only on predicting <code>is_spam</code>.
  </p>

</div>

### Train a Naive Bayes spam filter

In [11]:
# TODO: Import GaussianNB from sklearn.naive_bayes
# TODO: Create a GaussianNB model
# TODO: Train the model using X_train_spam and y_train_spam
# TODO: Predict on X_val_spam
# TODO: Evaluate the model using Accuracy and F1-score

from sklearn.naive_bayes import GaussianNB

model = GaussianNB()

model.fit(X_train,y_train)

y_pred_nb = model.predict(X_val)



In [12]:
evaluate_model(y_val,y_pred_nb,"Gaussian Naive Bayes")

 Performance Metrics: Gaussian Naive Bayes
------------------------------------------
Accuracy:  0.6908
Precision: 0.8359
Recall:    0.3247
F1-Score:  0.4677
------------------------------------------
Confusion Breakdown ====> TP: 163 | TN: 666 | FP: 32 | FN: 339


### Random Forest and GridSearchCV

In [13]:
# TODO: Import RandomForestClassifier
# TODO: Import GridSearchCV
# TODO: Create a RandomForestClassifier
# TODO: Create a param_grid for different max_depth values
# TODO: Use GridSearchCV with scoring="f1"
# TODO: Fit GridSearchCV on the training data
# TODO: Print the best parameters
# TODO: Predict using the best model
# TODO: Evaluate the Random Forest model

from sklearn.ensemble import RandomForestClassifier

from sklearn.model_selection import GridSearchCV

rf = RandomForestClassifier(random_state=2006)

param_grid = {
    'max_depth': [i for i in range(1,20)]
}

grid_search = GridSearchCV(
    estimator=rf,
    scoring="f1",
    param_grid=param_grid,
)

grid_search.fit(X_train,y_train)

best_rf_model = grid_search.best_estimator_
y_pred_rf = best_rf_model.predict(X_val)

print(f"Best parameters found{grid_search.best_params_}")

evaluate_model(y_val, y_pred_rf, model_name="Tuned Random Forest (Sklearn)")

Best parameters found{'max_depth': 6}
 Performance Metrics: Tuned Random Forest (Sklearn)
------------------------------------------
Accuracy:  0.6967
Precision: 0.7363
Recall:    0.4283
F1-Score:  0.5416
------------------------------------------
Confusion Breakdown ====> TP: 215 | TN: 621 | FP: 77 | FN: 287


### Final model comparison

In [14]:
# TODO: Create a DataFrame comparing:
#       - Scratch Decision Tree
#       - Naive Bayes
#       - Random Forest
#
# Columns:
#       - Model
#       - Accuracy
#       - F1-score
#
# TODO: Display the comparison table
from sklearn.metrics import accuracy_score,f1_score

def get_metrics_scratch(y_true, y_pred):
    y_true = np.array(y_true)
    y_pred = np.array(y_pred)
    
    tp = np.sum((y_true == 1) & (y_pred == 1))
    tn = np.sum((y_true == 0) & (y_pred == 0))
    fp = np.sum((y_true == 0) & (y_pred == 1))
    fn = np.sum((y_true == 1) & (y_pred == 0))

    accuracy = (tp + tn) / (tp + tn + fp + fn)
    precision = tp / (tp + fp)
    recall = tp / (fn + tp)
    f1_score = 2 * (precision * recall) / (precision + recall)
    
    return accuracy, f1_score

acc_tree, f1_tree = get_metrics_scratch(y_val, pred_spam)
acc_nb, f1_nb = get_metrics_scratch(y_val, y_pred_nb)
acc_rf, f1_rf = get_metrics_scratch(y_val, y_pred_rf)

comparison_data = {
    "Model": [
        "Scratch Decision Tree", 
        "Naive Bayes", 
        "Random Forest"
    ],
    "Accuracy": [acc_tree, acc_nb, acc_rf],
    "F1-score": [f1_tree, f1_nb, f1_rf]
}

df_comparison = pd.DataFrame(comparison_data)

print("$ Metrics calculated from scratch $")

display(df_comparison)


acc_tree_sk = accuracy_score(y_val, pred_spam)
f1_tree_sk = f1_score(y_val, pred_spam, average="binary")
acc_nb_sk = accuracy_score(y_val, y_pred_nb)
f1_nb_sk = f1_score(y_val, y_pred_nb, average="binary")
acc_rf_sk = accuracy_score(y_val, y_pred_rf)
f1_rf_sk = f1_score(y_val, y_pred_rf, average="binary")

comparison_data_sk = {
    "Model (Sklearn Metrics)": [
        "Scratch Decision Tree", 
        "Naive Bayes", 
        "Random Forest"
    ],
    "Accuracy": [acc_tree_sk, acc_nb_sk, acc_rf_sk],
    "F1-score": [f1_tree_sk, f1_nb_sk, f1_rf_sk]
}

df_comparison_sk = pd.DataFrame(comparison_data_sk)
print("$ Metrics Calculated with Sklearn $")
display(df_comparison_sk)



$ Metrics calculated from scratch $


,Model,Accuracy,F1-score
0,Scratch Decision Tree,0.697500,0.550186
1,Naive Bayes,0.690833,0.467719
2,Random Forest,0.696667,0.541562


$ Metrics Calculated with Sklearn $


,Model (Sklearn Metrics),Accuracy,F1-score
0,Scratch Decision Tree,0.697500,0.550186
1,Naive Bayes,0.690833,0.467719
2,Random Forest,0.696667,0.541562


***Gaussian Naive Bayes***

Accuracy (69.08%): The model correctly classifies messages as spam or healthy about 69% of the time.

Precision (83.59%): When Naive Bayes flags a message as "Spam," it is highly reliable—correct nearly 84% of the time.

Recall (32.47%): The model is extremely conservative. It only captures about 32.5% of the actual spam messages in the system.

F1-score (46.77%): The harmonic mean is heavily dragged down by the low recall rate.

***Random Forest***

Accuracy (69.67%): Overall accuracy is almost identical to Naive Bayes.

Precision (73.63%): Precision drops slightly compared to Naive Bayes, meaning it is a bit more prone to false alarms.

Recall (42.83%): It catches roughly 43% of the spam messages—a 10% improvement over Naive Bayes.

F1-score (54.16%): Shows a better overall balance between precision and recall than Naive Bayes.

***Questions:***

*Q1:*

No, this assumption is not valid. We cannot ignore the correlation coefficient because it is not zero; the urgent keywords and links are dependent features. As we can see from our results, the precision is high but the recall is significantly low, which happens precisely because of the Naive Bayes independence assumption for these features.

*Q2:*

The model becomes significantly simpler; however, this causes poor results in accuracy, precision, and other metrics due to underfitting. Because this model makes overly simplified decisions about the data, it indicates a high bias.

*Q3:*

Avoiding false negatives is much more important here. If we have low recall and a high false negative rate, it means that we miss a lot of lost messages and mistakenly mark them as healthy and well-received. On the other hand, if we have low precision, it means we classify a correctly received message as lost when we actually have it. Therefore, completely losing a message without the system noticing is much worse than mistakenly treating a successfully received message as lost.

<div style="
  background: linear-gradient(135deg, #9668b1, #8fb5d7, #78b9f6);
  padding: 30px;
  border-radius: 26px;
  text-align: center;
  color: #1d3557;
  box-shadow: 0 8px 22px rgba(0,0,0,0.16);
  margin-bottom: 24px;
">

  <h1 style="font-size: 42px; margin-bottom: 8px;">
    🧠 Mental Wellness AI Guardian 🌿
  </h1>

  <h2 style="font-size: 24px; margin-top: 0;">
    Build a Simple Neural Network from Scratch
  </h2>

  <p style="font-size: 18px; line-height: 1.7;">
    Welcome, wellness data scientist!  
    In this mission, you will train a small neural network to understand student habits
    and classify mental state patterns.
  </p>

  <div style="
    background-color: rgba(255,255,255,0.72);
    display: inline-block;
    padding: 10px 18px;
    border-radius: 18px;
    margin-top: 10px;
    font-size: 17px;
  ">
    💤 Sleep • 📰 News • 📱 Screen Time • 🔐 VPN Attempts • 🤖 Neural Network
  </div>

</div>

### Imports

In [15]:
import numpy as np

### Logistic regression from scratch

In [16]:
class SoftmaxRegressionScratch:
        """
        Multiclass logistic regression using softmax and gradient descent.
        """
        def __init__(self, learning_rate=0.01, n_iterations=1000):
            """
            Initialize the model.
            Args:
                learning_rate (float): step size for gradient descent.
                n_iterations (int): number of training epochs.
            """
            ## TODO: self.lr = ...
            ## TODO: self.n_iter = ...
            self.weights = None   
            self.bias = None     
            self.lr = learning_rate
            self.n_iter = n_iterations
            self.loss_history = []

        def _softmax(self, z):
            """
            Compute softmax probabilities for each sample.
            Args:
                z (np.ndarray): linear output, shape (n_samples, n_classes)
            Returns:
                np.ndarray: probabilities, shape (n_samples, n_classes)
            """
            # TODO: implement softmax with numerical stability
            exp_z = np.exp(z - np.max(z, axis = 1,keepdims=True))
            probabilites = exp_z / np.sum(exp_z,axis=1,keepdims=True)

            return probabilites

        def _one_hot(self, y, n_classes):
            """
            Convert integer labels to one-hot encoding.
            Args:
                y (np.ndarray): labels, shape (n_samples,)
                n_classes (int): number of unique classes
            Returns:
                np.ndarray: one-hot matrix, shape (n_samples, n_classes)
            """
            # TODO: implement one-hot encoding
            return np.eye(n_classes)[y]

        def fit(self, X, y):
            """
            Train the model using gradient descent.
            Args:
                X (np.ndarray): training features, shape (n_samples, n_features)
                y (np.ndarray): training labels, shape (n_samples,)
            """
            n_samples, n_features = X.shape
            n_classes = len(np.unique(y))
            
            # Initialize parameters
            # TODO: initialize self.weights (zeros) and self.bias (zeros)
            self.weights = np.zeros((n_features,n_classes))
            self.bias = np.zeros(n_classes)
            # One-hot encode labels
            # TODO: call self._one_hot(y, n_classes)
            one_hot = self._one_hot(y,n_classes)
            # Gradient descent loop
            for i in range(self.n_iter):
                # Forward pass: compute linear output z, then probabilities
                # TODO
                z = np.dot(X,self.weights) + self.bias
                probablities = self._softmax(z)
                # Compute cross-entropy loss (store in self.loss_history if desired)
                loss = -np.mean(np.sum(one_hot * np.log(probablities),axis=1))
                self.loss_history.append(loss)
                # Compute gradients
                # TODO: dw = 
                # TODO: db = 
                dz = probablities - one_hot
                dw = (1/n_samples) * np.dot(X.T,dz)
                db = (1/n_samples) * np.sum(dz,axis=0)
                # Update parameters
                # TODO: 
                self.weights -= self.lr*dw
                self.bias -= self.lr*db

                # print loss every 200 iterations
                if i % 200 == 0:
                    print(f"Iter {i}: loss = {loss:.4f}")

        def predict_proba(self, X):
            """
            Predict class probabilities for new data.
            Args:
                X (np.ndarray): features, shape (n_samples, n_features)
            Returns:
                np.ndarray: probabilities, shape (n_samples, n_classes)
            """
            # TODO: compute linear output and apply softmax
            z = np.dot(X, self.weights) + self.bias
            
            probabilities = self._softmax(z)
            
            return probabilities

        def predict(self, X):
            """
            Predict class labels for new data.
            Args:
                X (np.ndarray): features, shape (n_samples, n_features)
            Returns:
                np.ndarray: predicted labels, shape (n_samples,)
            """
            # TODO: get probabilities and return argmax
            probabilities  = self.predict_proba(X)

            predictions = np.argmax(probabilities,axis=1)

            return predictions

### Load data

In [17]:
## TODO:
df2 = pd.read_csv("student_burnout.csv")

print(df2.head())
print(df2.shape)

df2 = df2.dropna()

df2 = df2.drop_duplicates()

print(df2.shape)

   sleep_hours_avg  news_read_count  vpn_connection_attempts  pending_hw  \
0              5.9                5                        3           2   
1              8.8                4                        9           0   
2              7.7                2                        3           1   
3              7.0                5                        8           2   
4              4.8                8                        5           0   

   screen_time_hours  mental_state  
0                9.7             1  
1               12.6             0  
2                6.8             0  
3                2.8             1  
4                8.2             1  
(1000, 6)
(999, 6)


### Preprocessing

In [18]:
# TODO: Separate features (X) and target (y = mental_state)

# TODO: Split into train (80%) and validation (20%) with stratification

# TODO: Scale features using StandardScaler

X = df2.drop(columns=["mental_state"])
y = df2["mental_state"]

shuffled_indices_logistic_regression = np.random.permutation(len(X))
test_size_logistic_regression = 0.2

split_logistic_regression = int(len(X) * (1 - test_size_logistic_regression))



### Train and evaluate custom model

In [19]:
# custom_model = ...

# TODO: Measure training time and call fit()

# TODO: Predict on validation set and compute accuracy, classification report,
#       and confusion matrix

### Compare with sklearn's LogisticRegression

In [20]:
# TODO: Train sklearn's LogisticRegression (multinomial, lbfgs)
# TODO: Compare training time and performance metrics

### Imbalanced Data Analysis

In [21]:
# TODO: Print class distribution in training set
# TODO: Explain why accuracy alone is misleading
# TODO: Show macro/weighted F1 scores

<div style="
  background: linear-gradient(135deg, #7dbcf3, #9ff98d, #d6a870);
  padding: 30px;
  border-radius: 26px;
  text-align: center;
  color: #1d3557;
  box-shadow: 0 8px 20px rgba(0,0,0,0.15);
">

  <h1 style="font-size: 42px; margin-bottom: 8px;">
    🏔️ Payamnouria Mountain AI Mission 🤖
  </h1>

  <h2 style="font-size: 24px; margin-top: 0;">
    Predict Safe and Dangerous Trekking Routes
  </h2>

  <p style="font-size: 18px;">
    Welcome, young data explorer! Today you will help hikers choose safer mountain paths.
  </p>

</div>

### Imports

In [22]:

# TODO:

### Load the dataset

In [23]:
CSV_PATH = "trekking_expedition.csv"

# TODO: Read the CSV file into a DataFrame named df
# TODO: Print the shape of the dataset
# TODO: Display the first 5 rows

### Basic cleaning

In [24]:
required_columns = [
    "slope_angle",
    "wolf_prob",
    "rain_mm",
    "cold_resistance",
    "is_safe"
]

# TODO: Check that all required columns exist in df

# TODO: Remove duplicate rows

# TODO: Convert these columns to numeric:
#       slope_angle, wolf_prob, rain_mm, is_safe

# TODO: Remove rows with missing values

# TODO: Keep only rows where:
#       slope_angle > 0
#       0 <= wolf_prob <= 1
#       rain_mm >= 0
#       is_safe is either 0 or 1

# TODO: Keep only valid cold_resistance values:
#       Low, Medium, High

# TODO: Convert is_safe to integer

# TODO: Print the new shape of the cleaned dataset
# TODO: Print the target distribution


### Create environmental_danger

In [25]:
def add_environmental_danger(data):
    data = data.copy()

    # TODO: Create a new column named environmental_danger
    # Formula:
    # environmental_danger = slope_angle * rain_mm * (1 + wolf_prob)

    # Your code here:

    return data

### Preprocess Train and Validation

In [26]:
def preprocess_train_validation(X_train, X_val):
    X_train = X_train.copy()
    X_val = X_val.copy()

    categorical_columns = ["cold_resistance"]

    numeric_columns = [
        col for col in X_train.columns
        if col not in categorical_columns
    ]

    # -----------------------------------------------------
    # OneHotEncoder for cold_resistance
    # -----------------------------------------------------

    # TODO: Create OneHotEncoder with:
    #       handle_unknown="ignore"
    #       sparse_output=False

    # TODO: Fit encoder only on X_train[categorical_columns]

    # TODO: Transform both X_train and X_val categorical columns

    # TODO: Convert encoded arrays to DataFrames
    #       Use encoder.get_feature_names_out(categorical_columns)

    # -----------------------------------------------------
    # StandardScaler for numerical columns
    # -----------------------------------------------------

    # TODO: Create a StandardScaler

    # TODO: Fit scaler only on X_train[numeric_columns]

    # TODO: Transform both X_train and X_val numeric columns

    # TODO: Convert scaled arrays to DataFrames

    # -----------------------------------------------------
    # Combine numeric and categorical features
    # -----------------------------------------------------

    # TODO: Concatenate processed numeric and categorical DataFrames
    #       for both Train and Validation

    # TODO: Return X_train_processed and X_val_processed



### Train and evaluate model

In [27]:
def train_and_evaluate(data, experiment_name):
    print("\n" + "=" * 70)
    print(experiment_name)
    print("=" * 70)

    # TODO: Set X by dropping "is_safe"
    # TODO: Set y as data["is_safe"]

    # TODO: Split X and y into Train and Validation
    #       Use test_size=0.2, random_state=42, stratify=y

    # TODO: Preprocess Train and Validation using preprocess_train_validation

    # TODO: Create DecisionTreeClassifier
    #       Suggested parameters:
    #       max_depth=4
    #       min_samples_split=25
    #       min_samples_leaf=12
    #       random_state=42

    # TODO: Fit the model on processed Train data

    # TODO: Predict on processed Validation data

    # TODO: Calculate Accuracy and F1-score

    # TODO: Print Accuracy and F1-score

    # TODO: Return a dictionary with:
    #       experiment, accuracy, f1_score


### Run experiments

In [28]:
# TODO: Create df_baseline as a copy of df

# TODO: Train and evaluate the baseline model
#       Experiment name:
#       "Baseline Decision Tree - Without environmental_danger"

# TODO: Create df_engineered by calling add_environmental_danger(df)

# TODO: Train and evaluate the engineered model
#       Experiment name:
#       "Decision Tree - With environmental_danger"


### Final comparison

In [29]:
# TODO: Create a DataFrame from the two result dictionaries

# TODO: Print the comparison table rounded to 4 decimals

# TODO: Print the change in Accuracy

# TODO: Print the change in F1-score

<div style="
  background: linear-gradient(135deg, #f197a7, #efb66f, #f9fc95);
  padding: 30px;
  border-radius: 26px;
  text-align: center;
  color: #3a2c2c;
  box-shadow: 0 8px 22px rgba(0,0,0,0.16);
  margin-bottom: 24px;
">

  <h1 style="font-size: 42px; margin-bottom: 8px;">
    🎵 Smofy Koohestan AI Party 🎸
  </h1>

  <h2 style="font-size: 24px; margin-top: 0;">
    Build a KNN Music Genre Recommender
  </h2>

  <p style="font-size: 18px; line-height: 1.7;">
    Welcome, music data explorer!  
    In this mission, you will help students find their favorite music style using age,
    daily listening time, favorite instruments, and personality traits.
  </p>

  <div style="
    background-color: rgba(255,255,255,0.72);
    display: inline-block;
    padding: 10px 18px;
    border-radius: 18px;
    margin-top: 10px;
    font-size: 17px;
  ">
    🎧 Student Data • 📏 Distance Calculation • 👥 K-Nearest Neighbors • 🎶 Genre Prediction
  </div>

</div>

### Imports

In [30]:
## TODO:

### Data Preparation

In [31]:
# TODO: Load csv
# df = ...

# TODO: Separate features (X) and target (y = music_genre)

# TODO: Identify categorical columns and numeric columns 

# TODO: Apply one-hot encoding to categorical columns (use OneHotEncoder)
# TODO: Combine encoded categorical with numeric columns

# TODO: Split into train (80%) and validation (20%) with stratification

# TODO: Scale all features using StandardScaler (fit on train, transform both)


### Train and evaluate custom KNN (with k=5)

In [32]:
# TODO: Create KNeighborsClassifier model with k=5

# TODO: Fit the model on scaled training data

# TODO: Predict on validation data

# TODO: Compute accuracy and macro F1-score

# TODO: Print classification report



### Effect of k on bias/variance (plot)

In [33]:
k_values = [1, 3, 5, 7, 11, 21, 51]
train_accuracies = []
val_accuracies = []

for k in k_values:
    knn = KNeighborsClassifier(k=k)
    knn.fit(X_train_scaled, y_train)
    
    # Train accuracy (predict on training set)
    y_train_pred = knn.predict(X_train_scaled)
    train_acc = accuracy_score(y_train, y_train_pred)
    train_accuracies.append(train_acc)
    
    # Validation accuracy
    y_val_pred = knn.predict(X_val_scaled)
    val_acc = accuracy_score(y_val, y_val_pred)
    val_accuracies.append(val_acc)
    
    print(f"k={k:2d} -> Train Acc: {train_acc:.4f}, Val Acc: {val_acc:.4f}")


NameError: name 'KNeighborsClassifier' is not defined

In [ ]:
plt.figure(figsize=(8, 5))
plt.plot(k_values, train_accuracies, marker='o', label='Train Accuracy')
plt.plot(k_values, val_accuracies, marker='s', label='Validation Accuracy')
plt.xlabel('k (number of neighbors)')
plt.ylabel('Accuracy')
plt.title('Effect of k on Model Performance (KNN)')
plt.legend()
plt.grid(True)
plt.show()
